From c87755f74b4379079ae46dbcd736c5f03d3091f4 Mon Sep 17 00:00:00 2001
From: Daniel Kampert <DanielKampert@kampis-elektroecke.de>
Date: Thu, 26 Feb 2026 11:06:25 +0100
Subject: [PATCH] feat(esp_lcd): Add PSRAM mode support in SPI LCD panel IO

Add support for the `SPI_TRANS_DMA_USE_PSRAM` flag to the
`esp_lcd_panel_io_spi_transmit()` function, allowing it to
utilize PSRAM for DMA transfers when the flag is set.

Closes: #18283

Signed-off-by: Daniel Kampert <DanielKampert@kampis-elektroecke.de>
---
 components/esp_lcd/include/esp_lcd_io_spi.h   |  1 +
 components/esp_lcd/spi/esp_lcd_panel_io_spi.c | 19 +++++++++++++++++++
 2 files changed, 20 insertions(+)

diff --git a/components/esp_lcd/include/esp_lcd_io_spi.h b/components/esp_lcd/include/esp_lcd_io_spi.h
index 3a4b338b35..4871cecb40 100644
--- a/components/esp_lcd/include/esp_lcd_io_spi.h
+++ b/components/esp_lcd/include/esp_lcd_io_spi.h
@@ -41,6 +41,7 @@ typedef struct {
         unsigned int sio_mode: 1;        /*!< Read and write through a single data line (MOSI) */
         unsigned int lsb_first: 1;       /*!< transmit LSB bit first */
         unsigned int cs_high_active: 1;  /*!< CS line is high active */
+        unsigned int psram_mode: 1;      /*!< Enable direct transmissions from the PSRAM */
     } flags; /*!< Extra flags to fine-tune the SPI device */
 } esp_lcd_panel_io_spi_config_t;
 
diff --git a/components/esp_lcd/spi/esp_lcd_panel_io_spi.c b/components/esp_lcd/spi/esp_lcd_panel_io_spi.c
index eae42c890d..2063a8f005 100644
--- a/components/esp_lcd/spi/esp_lcd_panel_io_spi.c
+++ b/components/esp_lcd/spi/esp_lcd_panel_io_spi.c
@@ -63,6 +63,7 @@ typedef struct {
         unsigned int dc_param_level: 1;  // Indicates the level of DC line when transferring parameters
         unsigned int octal_mode: 1;      // Indicates whether the transmitting is enabled with octal mode (8 data lines)
         unsigned int quad_mode: 1;       // Indicates whether the transmitting is enabled with quad mode (4 data lines)
+        unsigned int psram_mode: 1;       // Indicates whether the transmitting is done from PSRAM
     } flags;
     lcd_spi_trans_descriptor_t trans_pool[]; // Transaction pool
 } esp_lcd_panel_io_spi_t;
@@ -106,6 +107,7 @@ esp_err_t esp_lcd_new_panel_io_spi(esp_lcd_spi_bus_handle_t bus, const esp_lcd_p
     spi_panel_io->flags.dc_param_level = !io_config->flags.dc_low_on_param;
     spi_panel_io->flags.octal_mode = io_config->flags.octal_mode;
     spi_panel_io->flags.quad_mode = io_config->flags.quad_mode;
+    spi_panel_io->flags.psram_mode = io_config->flags.psram_mode;
     spi_panel_io->on_color_trans_done = io_config->on_color_trans_done;
     spi_panel_io->user_ctx = io_config->user_ctx;
     spi_panel_io->lcd_cmd_bits = io_config->lcd_cmd_bits;
@@ -232,6 +234,10 @@ static esp_err_t panel_io_spi_tx_param(esp_lcd_panel_io_t *io, int lcd_cmd, cons
         lcd_trans->base.flags |= (SPI_TRANS_MULTILINE_CMD | SPI_TRANS_MULTILINE_ADDR | SPI_TRANS_MODE_OCT);
     }
 
+    if (spi_panel_io->flags.psram_mode) {
+        lcd_trans->base.flags |= SPI_TRANS_DMA_USE_PSRAM;
+    }
+
     if (send_cmd) {
         spi_lcd_prepare_cmd_buffer(spi_panel_io, &lcd_cmd);
         lcd_trans->flags.dc_gpio_level = spi_panel_io->flags.dc_cmd_level; // set D/C level in command phase
@@ -286,6 +292,10 @@ static esp_err_t panel_io_spi_rx_param(esp_lcd_panel_io_t *io, int lcd_cmd, void
         lcd_trans->base.flags |= (SPI_TRANS_MULTILINE_CMD | SPI_TRANS_MULTILINE_ADDR | SPI_TRANS_MODE_OCT);
     }
 
+    if (spi_panel_io->flags.psram_mode) {
+        lcd_trans->base.flags |= SPI_TRANS_DMA_USE_PSRAM;
+    }
+
     if (send_cmd) {
         spi_lcd_prepare_cmd_buffer(spi_panel_io, &lcd_cmd);
         lcd_trans->flags.dc_gpio_level = spi_panel_io->flags.dc_cmd_level; // set D/C level in command phase
@@ -347,6 +357,11 @@ static esp_err_t panel_io_spi_tx_color(esp_lcd_panel_io_t *io, int lcd_cmd, cons
             // use 8 lines for transmitting command, address and data
             lcd_trans->base.flags |= (SPI_TRANS_MULTILINE_CMD | SPI_TRANS_MULTILINE_ADDR | SPI_TRANS_MODE_OCT);
         }
+
+        if (spi_panel_io->flags.psram_mode) {
+            lcd_trans->base.flags |= SPI_TRANS_DMA_USE_PSRAM;
+        }
+
         // command is short, using polling mode
         ret = spi_device_polling_transmit(spi_panel_io->spi_dev, &lcd_trans->base);
         ESP_GOTO_ON_ERROR(ret, err, TAG, "spi transmit (polling) command failed");
@@ -391,6 +406,10 @@ static esp_err_t panel_io_spi_tx_color(esp_lcd_panel_io_t *io, int lcd_cmd, cons
             lcd_trans->base.flags |= SPI_TRANS_MODE_QIO;
         }
 
+        if (spi_panel_io->flags.psram_mode) {
+            lcd_trans->base.flags |= SPI_TRANS_DMA_USE_PSRAM;
+        }
+
         // color data is usually large, using queue+blocking mode
         ret = spi_device_queue_trans(spi_panel_io->spi_dev, &lcd_trans->base, portMAX_DELAY);
         ESP_GOTO_ON_ERROR(ret, err, TAG, "spi transmit (queue) color failed");
-- 
2.53.0.windows.1